src/python/pants/testutil/test_base.py

```diff
@@ -91,6 +121,40 @@ def request_product(
         result = assert_single_element(self.scheduler.product_request(product_type, [subject]))
         return cast(TestBase._P, result)
 
+    def run_goal_rule(
+        self,
+        goal: Type[Goal],
+        *,
+        global_args: Optional[Iterable[str]] = None,
+        args: Optional[Iterable[str]] = None,
+        env: Optional[Mapping[str, str]] = None,
+    ) -> GoalRuleResult:
+        options_bootstrapper = create_options_bootstrapper(
+            args=(*(global_args or []), goal.name, *(args or [])), env=env,
+        )
+
+        raw_specs = options_bootstrapper.get_full_options(
+            [*GlobalOptions.known_scope_infos(), *goal.subsystem_cls.known_scope_infos()]
+        ).specs
+        specs = SpecsParser(self.build_root).parse_specs(raw_specs)
+
+        stdout, stderr = StringIO(), StringIO()
+        console = Console(stdout=stdout, stderr=stderr)
+
+        exit_code = self.scheduler.run_goal_rule(
+            goal,
+            Params(
+                specs,
+                console,
+                options_bootstrapper,
+                Workspace(self.scheduler),
+                InteractiveRunner(self.scheduler),
+            ),
+        )
+
```

Comment:
The old code used to assert that the exit code was a certain value. I removed that to leave it up to call sites to do.